src/rdkafka_topic.c

```diff
@@ -603,17 +603,14 @@ int rd_kafka_toppar_broker_update(rd_kafka_toppar_t *rktp,
                                   int32_t broker_id,
                                   rd_kafka_broker_t *rkb,
                                   const char *reason) {
-
-        rktp->rktp_broker_id = broker_id;
-
         if (!rkb) {
                 int had_broker = rktp->rktp_broker ? 1 : 0;
                 rd_kafka_toppar_broker_delegate(rktp, NULL);
                 return had_broker ? -1 : 0;
         }
 
         if (rktp->rktp_broker) {
-                if (rktp->rktp_broker == rkb) {
+                if (rktp->rktp_broker == rkb && !rktp->rktp_next_broker) {
```

Comment:
Can you explain that change to me? This will make sure that we update the broker if a next one has been returned by the server?

Reply:
correct, if `rktp_next_broker` is set that means we're actively moving this partition from `rktp_broker` to `rktp_next_broker`.

Say we have two brokers for this partition: A is the leader and B is the follower. We're currently fetching from broker B, but we get a signal to switch back to broker A. We go through this code path and set `rktp_next_broker = A`. Then, before this first migration is processed, we get a metadata update that says broker B is now the leader and A is no longer in the ISR. We'll go through this same code path, see that `rktp_broker == B` and ignore that update - we'll process the migration to A and get stuck fetching from the wrong broker until the next leader epoch change.